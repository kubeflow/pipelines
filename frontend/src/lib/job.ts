import { ParameterValue } from '../lib/parameter';

export interface JobStep {
  name: string;
  start: number;
  end: number;
  state: string;
  outputs: string;
}

export interface Job {
  end: number;
  id: number;
  pipelineId: number;
  parameterValues: { [key: string]: ParameterValue };
  progress: number;
  recurring: boolean;
  recurringIntervalHours: number;
  start: number;
  state: 'not started' | 'running' | 'errored' | 'completed';
  steps: JobStep[];
}
