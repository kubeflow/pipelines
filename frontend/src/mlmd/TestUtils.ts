/*
 * Copyright 2018,2021 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Artifact, Value } from 'src/third_party/mlmd';
import { ArtifactCustomProperties, ArtifactProperties } from './Api';

export const doubleValue = (number: number) => {
  const value = new Value();
  value.setDoubleValue(number);
  return value;
};

export const intValue = (number: number) => {
  const value = new Value();
  value.setIntValue(number);
  return value;
};

export const stringValue = (string: String) => {
  const value = new Value();
  value.setStringValue(String(string));
  return value;
};

export const buildTestModel = () => {
  const model = new Artifact();
  model.setId(1);
  model.setTypeId(1);
  model.setUri('gs://my-bucket/mnist');
  model.getPropertiesMap().set(ArtifactProperties.NAME, stringValue('test model'));
  model.getPropertiesMap().set(ArtifactProperties.DESCRIPTION, stringValue('A really great model'));
  model.getPropertiesMap().set(ArtifactProperties.VERSION, stringValue('v1'));
  model
    .getPropertiesMap()
    .set(ArtifactProperties.CREATE_TIME, stringValue('2019-06-12T01:21:48.259263Z'));
  model
    .getPropertiesMap()
    .set(
      ArtifactProperties.ALL_META,
      stringValue(
        '{"hyperparameters": {"early_stop": true, ' +
          '"layers": [10, 3, 1], "learning_rate": 0.5}, ' +
          '"model_type": "neural network", ' +
          '"training_framework": {"name": "tensorflow", "version": "v1.0"}}',
      ),
    );
  model
    .getCustomPropertiesMap()
    .set(ArtifactCustomProperties.WORKSPACE, stringValue('workspace-1'));
  model.getCustomPropertiesMap().set(ArtifactCustomProperties.RUN, stringValue('1'));
  return model;
};

export const testModel = buildTestModel();
