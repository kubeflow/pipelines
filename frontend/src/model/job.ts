import { ParameterValue } from './parameter';

export interface JobStep {
  name: string;
  start: number;
  end: number;
  state: string;
  outputs: string;
}

export type JobStatus = 'Skipped' | 'Running' | 'Error' | 'Failed' | 'Succeeded';

export interface Job {
  createdAt: string;
  finishedAt: string;
  name: string;
  parameterValues: { [key: string]: ParameterValue };
  pipelineId: number;
  progress: number;
  recurring: boolean;
  recurringIntervalHours: number;
  startedAt: string;
  status: JobStatus;
  steps: JobStep[];
}
