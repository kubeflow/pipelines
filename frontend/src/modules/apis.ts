import Template from "./template";

/**
 * Gets a list of the pipeline templates defined on the backend.
 */
export async function getTemplates(): Promise<Template[]> {
  const response = await fetch('/_templates');
  const templates: Template[] = await response.json();
  return templates;
}

/**
 * Gets the details of a certain template given its id.
 */
export async function getTemplate(id: number): Promise<Template> {
  const response = await fetch(`/_templates?id=${id}`);
  const templates: Template[] = await response.json();
  return templates[0];
}
