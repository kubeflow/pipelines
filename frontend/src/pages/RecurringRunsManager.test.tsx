/*
 * Copyright 2018 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import TestUtils from '../TestUtils';
import { ListRequest, Apis } from '../lib/Apis';
import { shallow, ReactWrapper, ShallowWrapper } from 'enzyme';
import RecurringRunsManager, { RecurringRunListProps } from './RecurringRunsManager';
import { ApiJob, ApiResourceType } from '../apis/job';

describe('RecurringRunsManager', () => {
  class TestRecurringRunsManager extends RecurringRunsManager {
    public async _loadRuns(request: ListRequest): Promise<string> {
      return super._loadRuns(request);
    }
    public _setEnabledState(id: string, enabled: boolean): Promise<void> {
      return super._setEnabledState(id, enabled);
    }
  }

  let tree: ReactWrapper | ShallowWrapper;

  const updateDialogSpy = jest.fn();
  const updateSnackbarSpy = jest.fn();
  const listJobsSpy = jest.spyOn(Apis.jobServiceApi, 'listJobs');
  const enableJobSpy = jest.spyOn(Apis.jobServiceApi, 'enableJob');
  const disableJobSpy = jest.spyOn(Apis.jobServiceApi, 'disableJob');
  jest.spyOn(console, 'error').mockImplementation();

  const JOBS: ApiJob[] = [
    {
      created_at: new Date(2018, 10, 9, 8, 7, 6),
      enabled: true,
      id: 'job1',
      name: 'test recurring run name',
    },
    {
      created_at: new Date(2018, 10, 9, 8, 7, 6),
      enabled: false,
      id: 'job2',
      name: 'test recurring run name2',
    },
    {
      created_at: new Date(2018, 10, 9, 8, 7, 6),
      id: 'job3',
      name: 'test recurring run name3',
    },
  ];

  function generateProps(): RecurringRunListProps {
    return {
      experimentId: 'test-experiment',
      history: {} as any,
      location: '' as any,
      match: {} as any,
      updateDialog: updateDialogSpy,
      updateSnackbar: updateSnackbarSpy,
    };
  }

  beforeEach(() => {
    listJobsSpy.mockReset();
    listJobsSpy.mockImplementation(() => ({ jobs: JOBS }));
    enableJobSpy.mockReset();
    disableJobSpy.mockReset();
    updateDialogSpy.mockReset();
    updateSnackbarSpy.mockReset();
  });

  afterEach(() => tree.unmount());

  it('calls API to load recurring runs', async () => {
    tree = shallow(<TestRecurringRunsManager {...generateProps()} />);
    await (tree.instance() as TestRecurringRunsManager)._loadRuns({});
    expect(listJobsSpy).toHaveBeenCalledTimes(1);
    expect(listJobsSpy).toHaveBeenLastCalledWith(
      undefined,
      undefined,
      undefined,
      ApiResourceType.EXPERIMENT,
      'test-experiment',
      undefined,
    );
    expect(tree.state('runs')).toEqual(JOBS);
    expect(tree).toMatchSnapshot();
  });

  it('shows error dialog if listing fails', async () => {
    TestUtils.makeErrorResponseOnce(listJobsSpy, 'woops!');
    jest.spyOn(console, 'error').mockImplementation();
    tree = shallow(<TestRecurringRunsManager {...generateProps()} />);
    await (tree.instance() as TestRecurringRunsManager)._loadRuns({});
    expect(listJobsSpy).toHaveBeenCalledTimes(1);
    expect(updateDialogSpy).toHaveBeenLastCalledWith(
      expect.objectContaining({
        content: 'List recurring run configs request failed with:\nwoops!',
        title: 'Error retrieving recurring run configs',
      }),
    );
    expect(tree.state('runs')).toEqual([]);
  });

  it('calls API to enable run', async () => {
    tree = shallow(<TestRecurringRunsManager {...generateProps()} />);
    await (tree.instance() as TestRecurringRunsManager)._setEnabledState('test-run', true);
    expect(enableJobSpy).toHaveBeenCalledTimes(1);
    expect(enableJobSpy).toHaveBeenLastCalledWith('test-run');
  });

  it('calls API to disable run', async () => {
    tree = shallow(<TestRecurringRunsManager {...generateProps()} />);
    await (tree.instance() as TestRecurringRunsManager)._setEnabledState('test-run', false);
    expect(disableJobSpy).toHaveBeenCalledTimes(1);
    expect(disableJobSpy).toHaveBeenLastCalledWith('test-run');
  });

  it('shows error if enable API call fails', async () => {
    tree = shallow(<TestRecurringRunsManager {...generateProps()} />);
    TestUtils.makeErrorResponseOnce(enableJobSpy, 'cannot enable');
    await (tree.instance() as TestRecurringRunsManager)._setEnabledState('test-run', true);
    expect(updateDialogSpy).toHaveBeenCalledTimes(1);
    expect(updateDialogSpy).toHaveBeenLastCalledWith(
      expect.objectContaining({
        content: 'Error changing enabled state of recurring run:\ncannot enable',
        title: 'Error',
      }),
    );
  });

  it('shows error if disable API call fails', async () => {
    tree = shallow(<TestRecurringRunsManager {...generateProps()} />);
    TestUtils.makeErrorResponseOnce(disableJobSpy, 'cannot disable');
    await (tree.instance() as TestRecurringRunsManager)._setEnabledState('test-run', false);
    expect(updateDialogSpy).toHaveBeenCalledTimes(1);
    expect(updateDialogSpy).toHaveBeenLastCalledWith(
      expect.objectContaining({
        content: 'Error changing enabled state of recurring run:\ncannot disable',
        title: 'Error',
      }),
    );
  });

  it('renders run name as link to its details page', () => {
    tree = TestUtils.mountWithRouter(<RecurringRunsManager {...generateProps()} />);
    expect(
      (tree.instance() as RecurringRunsManager)._nameCustomRenderer({
        id: 'run-id',
        value: 'test-run',
      }),
    ).toMatchSnapshot();
  });

  it('renders a disable button if the run is enabled, clicking the button calls disable API', async () => {
    tree = TestUtils.mountWithRouter(<RecurringRunsManager {...generateProps()} />);
    await TestUtils.flushPromises();
    tree.update();

    const enableBtn = tree.find('.tableRow Button').at(0);
    expect(enableBtn).toMatchSnapshot();

    enableBtn.simulate('click');
    await TestUtils.flushPromises();
    expect(disableJobSpy).toHaveBeenCalledTimes(1);
    expect(disableJobSpy).toHaveBeenLastCalledWith(JOBS[0].id);
  });

  it('renders an enable button if the run is disabled, clicking the button calls enable API', async () => {
    tree = TestUtils.mountWithRouter(<RecurringRunsManager {...generateProps()} />);
    await TestUtils.flushPromises();
    tree.update();

    const enableBtn = tree.find('.tableRow Button').at(1);
    expect(enableBtn).toMatchSnapshot();

    enableBtn.simulate('click');
    await TestUtils.flushPromises();
    expect(enableJobSpy).toHaveBeenCalledTimes(1);
    expect(enableJobSpy).toHaveBeenLastCalledWith(JOBS[1].id);
  });

  it("renders an enable button if the run's enabled field is undefined, clicking the button calls enable API", async () => {
    tree = TestUtils.mountWithRouter(<RecurringRunsManager {...generateProps()} />);
    await TestUtils.flushPromises();
    tree.update();

    const enableBtn = tree.find('.tableRow Button').at(2);
    expect(enableBtn).toMatchSnapshot();

    enableBtn.simulate('click');
    await TestUtils.flushPromises();
    expect(enableJobSpy).toHaveBeenCalledTimes(1);
    expect(enableJobSpy).toHaveBeenLastCalledWith(JOBS[2].id);
  });

  it('reloads the list of runs after enable/disabling', async () => {
    tree = TestUtils.mountWithRouter(<RecurringRunsManager {...generateProps()} />);
    await TestUtils.flushPromises();
    tree.update();

    const enableBtn = tree.find('.tableRow Button').at(0);
    expect(enableBtn).toMatchSnapshot();

    expect(listJobsSpy).toHaveBeenCalledTimes(1);
    enableBtn.simulate('click');
    await TestUtils.flushPromises();
    expect(listJobsSpy).toHaveBeenCalledTimes(2);
  });
});
