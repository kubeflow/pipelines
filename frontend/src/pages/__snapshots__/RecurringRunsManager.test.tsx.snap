// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecurringRunsManager calls API to load recurring runs 1`] = `
<Fragment>
  <Toolbar
    actions={Object {}}
    breadcrumbs={Array []}
    pageTitle="Recurring runs"
  />
  <CustomTable
    columns={
      Array [
        Object {
          "customRenderer": [Function],
          "flex": 2,
          "label": "Run name",
          "sortKey": "name",
        },
        Object {
          "flex": 2,
          "label": "Created at",
          "sortKey": "created_at",
        },
        Object {
          "customRenderer": [Function],
          "flex": 1,
          "label": "",
        },
      ]
    }
    disableSelection={true}
    emptyMessage="No recurring runs found in this experiment."
    filterLabel="Filter recurring runs"
    initialSortColumn="created_at"
    reload={[Function]}
    rows={
      Array [
        Object {
          "error": undefined,
          "id": "job1",
          "otherFields": Array [
            "test recurring run name",
            "11/9/2018, 8:07:06 AM",
            true,
          ],
        },
        Object {
          "error": undefined,
          "id": "job2",
          "otherFields": Array [
            "test recurring run name2",
            "11/9/2018, 8:07:06 AM",
            false,
          ],
        },
        Object {
          "error": undefined,
          "id": "job3",
          "otherFields": Array [
            "test recurring run name3",
            "11/9/2018, 8:07:06 AM",
            undefined,
          ],
        },
      ]
    }
    selectedIds={Array []}
    updateSelection={[Function]}
  />
</Fragment>
`;

exports[`RecurringRunsManager reloads the list of runs after enable/disabling 1`] = `
<Button
  className="root"
  classes={
    Object {
      "colorInherit": "MuiButton-colorInherit-836",
      "contained": "MuiButton-contained-826",
      "containedPrimary": "MuiButton-containedPrimary-827",
      "containedSecondary": "MuiButton-containedSecondary-828",
      "disabled": "MuiButton-disabled-835",
      "extendedFab": "MuiButton-extendedFab-833",
      "fab": "MuiButton-fab-832",
      "flat": "MuiButton-flat-820",
      "flatPrimary": "MuiButton-flatPrimary-821",
      "flatSecondary": "MuiButton-flatSecondary-822",
      "focusVisible": "MuiButton-focusVisible-834",
      "fullWidth": "MuiButton-fullWidth-840",
      "label": "MuiButton-label-816",
      "mini": "MuiButton-mini-837",
      "outlined": "MuiButton-outlined-823",
      "outlinedPrimary": "MuiButton-outlinedPrimary-824",
      "outlinedSecondary": "MuiButton-outlinedSecondary-825",
      "raised": "MuiButton-raised-829",
      "raisedPrimary": "MuiButton-raisedPrimary-830",
      "raisedSecondary": "MuiButton-raisedSecondary-831",
      "root": "MuiButton-root-815",
      "sizeLarge": "MuiButton-sizeLarge-839",
      "sizeSmall": "MuiButton-sizeSmall-838",
      "text": "MuiButton-text-817",
      "textPrimary": "MuiButton-textPrimary-818",
      "textSecondary": "MuiButton-textSecondary-819",
    }
  }
  color="primary"
  component="button"
  disableFocusRipple={false}
  disabled={false}
  fullWidth={false}
  mini={false}
  onClick={[Function]}
  size="medium"
  type="button"
  variant="text"
>
  <WithStyles(ButtonBase)
    className="MuiButton-root-815 MuiButton-text-817 MuiButton-textPrimary-818 MuiButton-flat-820 MuiButton-flatPrimary-821 root"
    component="button"
    disabled={false}
    focusRipple={true}
    focusVisibleClassName="MuiButton-focusVisible-834"
    onClick={[Function]}
    type="button"
  >
    <ButtonBase
      centerRipple={false}
      className="MuiButton-root-815 MuiButton-text-817 MuiButton-textPrimary-818 MuiButton-flat-820 MuiButton-flatPrimary-821 root"
      classes={
        Object {
          "disabled": "MuiButtonBase-disabled-766",
          "focusVisible": "MuiButtonBase-focusVisible-767",
          "root": "MuiButtonBase-root-765",
        }
      }
      component="button"
      disableRipple={false}
      disableTouchRipple={false}
      disabled={false}
      focusRipple={true}
      focusVisibleClassName="MuiButton-focusVisible-834"
      onClick={[Function]}
      tabIndex="0"
      type="button"
    >
      <button
        className="MuiButtonBase-root-765 MuiButton-root-815 MuiButton-text-817 MuiButton-textPrimary-818 MuiButton-flat-820 MuiButton-flatPrimary-821 root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-816"
        >
          <span>
            Enabled
          </span>
        </span>
        <NoSsr
          defer={false}
          fallback={null}
        >
          <WithStyles(TouchRipple)
            center={false}
            innerRef={[Function]}
          >
            <TouchRipple
              center={false}
              classes={
                Object {
                  "child": "MuiTouchRipple-child-812",
                  "childLeaving": "MuiTouchRipple-childLeaving-813",
                  "childPulsate": "MuiTouchRipple-childPulsate-814",
                  "ripple": "MuiTouchRipple-ripple-809",
                  "ripplePulsate": "MuiTouchRipple-ripplePulsate-811",
                  "rippleVisible": "MuiTouchRipple-rippleVisible-810",
                  "root": "MuiTouchRipple-root-808",
                }
              }
            >
              <TransitionGroup
                childFactory={[Function]}
                className="MuiTouchRipple-root-808"
                component="span"
                enter={true}
                exit={true}
              >
                <span
                  className="MuiTouchRipple-root-808"
                />
              </TransitionGroup>
            </TouchRipple>
          </WithStyles(TouchRipple)>
        </NoSsr>
      </button>
    </ButtonBase>
  </WithStyles(ButtonBase)>
</Button>
`;

exports[`RecurringRunsManager renders a disable button if the run is enabled, clicking the button calls disable API 1`] = `
<Button
  className="root"
  classes={
    Object {
      "colorInherit": "MuiButton-colorInherit-308",
      "contained": "MuiButton-contained-298",
      "containedPrimary": "MuiButton-containedPrimary-299",
      "containedSecondary": "MuiButton-containedSecondary-300",
      "disabled": "MuiButton-disabled-307",
      "extendedFab": "MuiButton-extendedFab-305",
      "fab": "MuiButton-fab-304",
      "flat": "MuiButton-flat-292",
      "flatPrimary": "MuiButton-flatPrimary-293",
      "flatSecondary": "MuiButton-flatSecondary-294",
      "focusVisible": "MuiButton-focusVisible-306",
      "fullWidth": "MuiButton-fullWidth-312",
      "label": "MuiButton-label-288",
      "mini": "MuiButton-mini-309",
      "outlined": "MuiButton-outlined-295",
      "outlinedPrimary": "MuiButton-outlinedPrimary-296",
      "outlinedSecondary": "MuiButton-outlinedSecondary-297",
      "raised": "MuiButton-raised-301",
      "raisedPrimary": "MuiButton-raisedPrimary-302",
      "raisedSecondary": "MuiButton-raisedSecondary-303",
      "root": "MuiButton-root-287",
      "sizeLarge": "MuiButton-sizeLarge-311",
      "sizeSmall": "MuiButton-sizeSmall-310",
      "text": "MuiButton-text-289",
      "textPrimary": "MuiButton-textPrimary-290",
      "textSecondary": "MuiButton-textSecondary-291",
    }
  }
  color="primary"
  component="button"
  disableFocusRipple={false}
  disabled={false}
  fullWidth={false}
  mini={false}
  onClick={[Function]}
  size="medium"
  type="button"
  variant="text"
>
  <WithStyles(ButtonBase)
    className="MuiButton-root-287 MuiButton-text-289 MuiButton-textPrimary-290 MuiButton-flat-292 MuiButton-flatPrimary-293 root"
    component="button"
    disabled={false}
    focusRipple={true}
    focusVisibleClassName="MuiButton-focusVisible-306"
    onClick={[Function]}
    type="button"
  >
    <ButtonBase
      centerRipple={false}
      className="MuiButton-root-287 MuiButton-text-289 MuiButton-textPrimary-290 MuiButton-flat-292 MuiButton-flatPrimary-293 root"
      classes={
        Object {
          "disabled": "MuiButtonBase-disabled-238",
          "focusVisible": "MuiButtonBase-focusVisible-239",
          "root": "MuiButtonBase-root-237",
        }
      }
      component="button"
      disableRipple={false}
      disableTouchRipple={false}
      disabled={false}
      focusRipple={true}
      focusVisibleClassName="MuiButton-focusVisible-306"
      onClick={[Function]}
      tabIndex="0"
      type="button"
    >
      <button
        className="MuiButtonBase-root-237 MuiButton-root-287 MuiButton-text-289 MuiButton-textPrimary-290 MuiButton-flat-292 MuiButton-flatPrimary-293 root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-288"
        >
          <span>
            Enabled
          </span>
        </span>
        <NoSsr
          defer={false}
          fallback={null}
        >
          <WithStyles(TouchRipple)
            center={false}
            innerRef={[Function]}
          >
            <TouchRipple
              center={false}
              classes={
                Object {
                  "child": "MuiTouchRipple-child-284",
                  "childLeaving": "MuiTouchRipple-childLeaving-285",
                  "childPulsate": "MuiTouchRipple-childPulsate-286",
                  "ripple": "MuiTouchRipple-ripple-281",
                  "ripplePulsate": "MuiTouchRipple-ripplePulsate-283",
                  "rippleVisible": "MuiTouchRipple-rippleVisible-282",
                  "root": "MuiTouchRipple-root-280",
                }
              }
            >
              <TransitionGroup
                childFactory={[Function]}
                className="MuiTouchRipple-root-280"
                component="span"
                enter={true}
                exit={true}
              >
                <span
                  className="MuiTouchRipple-root-280"
                />
              </TransitionGroup>
            </TouchRipple>
          </WithStyles(TouchRipple)>
        </NoSsr>
      </button>
    </ButtonBase>
  </WithStyles(ButtonBase)>
</Button>
`;

exports[`RecurringRunsManager renders an enable button if the run is disabled, clicking the button calls enable API 1`] = `
<Button
  className="root"
  classes={
    Object {
      "colorInherit": "MuiButton-colorInherit-484",
      "contained": "MuiButton-contained-474",
      "containedPrimary": "MuiButton-containedPrimary-475",
      "containedSecondary": "MuiButton-containedSecondary-476",
      "disabled": "MuiButton-disabled-483",
      "extendedFab": "MuiButton-extendedFab-481",
      "fab": "MuiButton-fab-480",
      "flat": "MuiButton-flat-468",
      "flatPrimary": "MuiButton-flatPrimary-469",
      "flatSecondary": "MuiButton-flatSecondary-470",
      "focusVisible": "MuiButton-focusVisible-482",
      "fullWidth": "MuiButton-fullWidth-488",
      "label": "MuiButton-label-464",
      "mini": "MuiButton-mini-485",
      "outlined": "MuiButton-outlined-471",
      "outlinedPrimary": "MuiButton-outlinedPrimary-472",
      "outlinedSecondary": "MuiButton-outlinedSecondary-473",
      "raised": "MuiButton-raised-477",
      "raisedPrimary": "MuiButton-raisedPrimary-478",
      "raisedSecondary": "MuiButton-raisedSecondary-479",
      "root": "MuiButton-root-463",
      "sizeLarge": "MuiButton-sizeLarge-487",
      "sizeSmall": "MuiButton-sizeSmall-486",
      "text": "MuiButton-text-465",
      "textPrimary": "MuiButton-textPrimary-466",
      "textSecondary": "MuiButton-textSecondary-467",
    }
  }
  color="secondary"
  component="button"
  disableFocusRipple={false}
  disabled={false}
  fullWidth={false}
  mini={false}
  onClick={[Function]}
  size="medium"
  type="button"
  variant="text"
>
  <WithStyles(ButtonBase)
    className="MuiButton-root-463 MuiButton-text-465 MuiButton-textSecondary-467 MuiButton-flat-468 MuiButton-flatSecondary-470 root"
    component="button"
    disabled={false}
    focusRipple={true}
    focusVisibleClassName="MuiButton-focusVisible-482"
    onClick={[Function]}
    type="button"
  >
    <ButtonBase
      centerRipple={false}
      className="MuiButton-root-463 MuiButton-text-465 MuiButton-textSecondary-467 MuiButton-flat-468 MuiButton-flatSecondary-470 root"
      classes={
        Object {
          "disabled": "MuiButtonBase-disabled-414",
          "focusVisible": "MuiButtonBase-focusVisible-415",
          "root": "MuiButtonBase-root-413",
        }
      }
      component="button"
      disableRipple={false}
      disableTouchRipple={false}
      disabled={false}
      focusRipple={true}
      focusVisibleClassName="MuiButton-focusVisible-482"
      onClick={[Function]}
      tabIndex="0"
      type="button"
    >
      <button
        className="MuiButtonBase-root-413 MuiButton-root-463 MuiButton-text-465 MuiButton-textSecondary-467 MuiButton-flat-468 MuiButton-flatSecondary-470 root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-464"
        >
          <span>
            Disabled
          </span>
        </span>
        <NoSsr
          defer={false}
          fallback={null}
        >
          <WithStyles(TouchRipple)
            center={false}
            innerRef={[Function]}
          >
            <TouchRipple
              center={false}
              classes={
                Object {
                  "child": "MuiTouchRipple-child-460",
                  "childLeaving": "MuiTouchRipple-childLeaving-461",
                  "childPulsate": "MuiTouchRipple-childPulsate-462",
                  "ripple": "MuiTouchRipple-ripple-457",
                  "ripplePulsate": "MuiTouchRipple-ripplePulsate-459",
                  "rippleVisible": "MuiTouchRipple-rippleVisible-458",
                  "root": "MuiTouchRipple-root-456",
                }
              }
            >
              <TransitionGroup
                childFactory={[Function]}
                className="MuiTouchRipple-root-456"
                component="span"
                enter={true}
                exit={true}
              >
                <span
                  className="MuiTouchRipple-root-456"
                />
              </TransitionGroup>
            </TouchRipple>
          </WithStyles(TouchRipple)>
        </NoSsr>
      </button>
    </ButtonBase>
  </WithStyles(ButtonBase)>
</Button>
`;

exports[`RecurringRunsManager renders an enable button if the run's enabled field is undefined, clicking the button calls enable API 1`] = `
<Button
  className="root"
  classes={
    Object {
      "colorInherit": "MuiButton-colorInherit-660",
      "contained": "MuiButton-contained-650",
      "containedPrimary": "MuiButton-containedPrimary-651",
      "containedSecondary": "MuiButton-containedSecondary-652",
      "disabled": "MuiButton-disabled-659",
      "extendedFab": "MuiButton-extendedFab-657",
      "fab": "MuiButton-fab-656",
      "flat": "MuiButton-flat-644",
      "flatPrimary": "MuiButton-flatPrimary-645",
      "flatSecondary": "MuiButton-flatSecondary-646",
      "focusVisible": "MuiButton-focusVisible-658",
      "fullWidth": "MuiButton-fullWidth-664",
      "label": "MuiButton-label-640",
      "mini": "MuiButton-mini-661",
      "outlined": "MuiButton-outlined-647",
      "outlinedPrimary": "MuiButton-outlinedPrimary-648",
      "outlinedSecondary": "MuiButton-outlinedSecondary-649",
      "raised": "MuiButton-raised-653",
      "raisedPrimary": "MuiButton-raisedPrimary-654",
      "raisedSecondary": "MuiButton-raisedSecondary-655",
      "root": "MuiButton-root-639",
      "sizeLarge": "MuiButton-sizeLarge-663",
      "sizeSmall": "MuiButton-sizeSmall-662",
      "text": "MuiButton-text-641",
      "textPrimary": "MuiButton-textPrimary-642",
      "textSecondary": "MuiButton-textSecondary-643",
    }
  }
  color="secondary"
  component="button"
  disableFocusRipple={false}
  disabled={false}
  fullWidth={false}
  mini={false}
  onClick={[Function]}
  size="medium"
  type="button"
  variant="text"
>
  <WithStyles(ButtonBase)
    className="MuiButton-root-639 MuiButton-text-641 MuiButton-textSecondary-643 MuiButton-flat-644 MuiButton-flatSecondary-646 root"
    component="button"
    disabled={false}
    focusRipple={true}
    focusVisibleClassName="MuiButton-focusVisible-658"
    onClick={[Function]}
    type="button"
  >
    <ButtonBase
      centerRipple={false}
      className="MuiButton-root-639 MuiButton-text-641 MuiButton-textSecondary-643 MuiButton-flat-644 MuiButton-flatSecondary-646 root"
      classes={
        Object {
          "disabled": "MuiButtonBase-disabled-590",
          "focusVisible": "MuiButtonBase-focusVisible-591",
          "root": "MuiButtonBase-root-589",
        }
      }
      component="button"
      disableRipple={false}
      disableTouchRipple={false}
      disabled={false}
      focusRipple={true}
      focusVisibleClassName="MuiButton-focusVisible-658"
      onClick={[Function]}
      tabIndex="0"
      type="button"
    >
      <button
        className="MuiButtonBase-root-589 MuiButton-root-639 MuiButton-text-641 MuiButton-textSecondary-643 MuiButton-flat-644 MuiButton-flatSecondary-646 root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-640"
        >
          <span>
            Disabled
          </span>
        </span>
        <NoSsr
          defer={false}
          fallback={null}
        >
          <WithStyles(TouchRipple)
            center={false}
            innerRef={[Function]}
          >
            <TouchRipple
              center={false}
              classes={
                Object {
                  "child": "MuiTouchRipple-child-636",
                  "childLeaving": "MuiTouchRipple-childLeaving-637",
                  "childPulsate": "MuiTouchRipple-childPulsate-638",
                  "ripple": "MuiTouchRipple-ripple-633",
                  "ripplePulsate": "MuiTouchRipple-ripplePulsate-635",
                  "rippleVisible": "MuiTouchRipple-rippleVisible-634",
                  "root": "MuiTouchRipple-root-632",
                }
              }
            >
              <TransitionGroup
                childFactory={[Function]}
                className="MuiTouchRipple-root-632"
                component="span"
                enter={true}
                exit={true}
              >
                <span
                  className="MuiTouchRipple-root-632"
                />
              </TransitionGroup>
            </TouchRipple>
          </WithStyles(TouchRipple)>
        </NoSsr>
      </button>
    </ButtonBase>
  </WithStyles(ButtonBase)>
</Button>
`;

exports[`RecurringRunsManager renders run name as link to its details page 1`] = `
<Link
  className="link"
  replace={false}
  to="/recurringrun/details/run-id"
>
  test-run
</Link>
`;
